noisy-simulator/src/operation.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! This module contains the `Operation` struct and the `operation!` macro
+//! to conveniently construct operations from Kraus matrices.
+
+#[cfg(test)]
+mod tests;
+use crate::SquareMatrix;
+
+/// A helper macro to write operations more conveniently.
+///
+/// Example usage:
+/// ```
+/// // Create operation from two 2x2 Kraus matrices.
+/// use noisy_simulator::{Operation, operation};
+///
+/// let op = operation!(
+///     [1., 0.;
+///      0., 0.;],
+///     [0., 0.;
+///      0., 0.;]
+/// );
+/// ```
+#[macro_export]
+macro_rules! operation {
+    ($([$($($v:expr),* );*]),*) => {
+        Operation::new(vec![
+            $(nalgebra::dmatrix![
+                $($(num_complex::Complex::<f64>::from($v)),* );*
+            ]),*
+        ])
+    };
+}
+
+/// This struct represents a quantum operation.
+/// A quantum operation is a linear transformation that maps a valid density
+/// matrix to another valid density matrices.
+#[derive(Clone)]
+pub struct Operation {
+    number_of_qubits: usize,
+    kraus_operators: Vec<SquareMatrix>,
+    operation_matrix: SquareMatrix,
+    effect_matrix: SquareMatrix,
+    effect_matrix_transpose: SquareMatrix,
+}
+
+impl Operation {
+    /// Construct an operation from a list of Kraus operators.
+    /// Matrices must be of dimension 2^k x 2^k, where k is an integer.
+    /// Returns `None` if the kraus matrices are ill formed.
+    pub fn new(kraus_operators: Vec<SquareMatrix>) -> Self {
+        let (dim, _) = kraus_operators
+            .first()
+            .expect("there should be at least one Kraus Operator")
+            .shape();
+
+        let number_of_qubits = (dim as f64).log2() as usize;
+        assert!(
+            1 << number_of_qubits == dim,
+            "kraus operators should have dimensions 2^k x 2^k"
+        );
```

Comment:
From an API perspective, do we want to crash the Python process when bad data is supplied? Should this return `Result<Self, crate::Error>` instead?

Reply:
Yes, nice catch. That was my intention when returning Result from the methods in both simulators but forgot to do the same for `Operation`.